cps/kni.c

```diff
@@ -404,6 +439,141 @@ inet6_print(const char *msg, struct in6_addr *in6)
 		inet_ntop(AF_INET6, &in6->s6_addr, buf, sizeof(buf)));
 }
 
+static int
+new_route(struct route_update *update, struct cps_config *cps_conf)
+{
+	int ret;
+	uint16_t proto;
+	char ip_buf[INET6_ADDRSTRLEN];
+	char ipp_buf[INET6_ADDRSTRLEN + 4];
+	int gw_fib_id;
+	struct ip_prefix prefix_info;
+	struct ipaddr gw_addr;
+	struct gk_fib *gw_fib;
+	struct gk_lpm *ltbl = &cps_conf->gk->lpm_tbl;
+
+	if (update->family == AF_INET) {
+		proto = ETHER_TYPE_IPv4;
+		gw_fib_id = lpm_lookup_ipv4(ltbl->lpm, update->gw.v4.s_addr);
+		if (gw_fib_id < 0)
+			return -1;
+		gw_fib = &ltbl->fib_tbl[gw_fib_id];
+
+		if (inet_ntop(AF_INET, &update->ip.v4.s_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+	} else if (likely(update->family == AF_INET6)) {
+		proto = ETHER_TYPE_IPv6;
+		gw_fib_id = lpm_lookup_ipv6(ltbl->lpm6, update->gw.v6.s6_addr);
+		if (gw_fib_id < 0)
+			return -1;
+		gw_fib = &ltbl->fib_tbl6[gw_fib_id];
+
+		if (inet_ntop(AF_INET6, &update->ip.v6.s6_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+	} else {
+		CPS_LOG(WARNING,
+			"cps update: unknown address family %d at %s\n",
+			update->family, __func__);
+		return -1;
+	}
+
+	prefix_info.str = ipp_buf;
+	prefix_info.addr.proto = proto;
+	rte_memcpy(&prefix_info.addr.ip, &update->ip,
+		sizeof(prefix_info.addr.ip));
+	prefix_info.len = update->prefix_len;
+
+	gw_addr.proto = proto;
+	rte_memcpy(&gw_addr.ip, &update->gw, sizeof(gw_addr.ip));
+
+	if (gw_fib->action == GK_FWD_NEIGHBOR_FRONT_NET) {
+		if (update->oif_index != 0) {
+			if (update->oif_index != cps_conf->front_kni_index) {
+				CPS_LOG(WARNING,
+					"The output KNI interface for prefix %s is not the front interface while the gateway for the prefix in Gatekeeper is a neighbor of the front network\n",
+					prefix_info.str);
+				return -1;
+			}
+		}
+
+		return add_fib_entry_numerical(&prefix_info, NULL, &gw_addr,
+			GK_FWD_GATEWAY_FRONT_NET, update->rt_proto,
+			cps_conf->gk);
+	}
+
+	if (gw_fib->action == GK_FWD_NEIGHBOR_BACK_NET) {
+		if (update->oif_index != 0) {
+			if (update->oif_index != cps_conf->back_kni_index) {
+				CPS_LOG(WARNING,
+					"The output KNI interface for prefix %s is not the back interface while the gateway for the prefix in Gatekeeper is a neighbor of the back network\n",
+					prefix_info.str);
+				return -1;
+			}
+		}
+
+		return add_fib_entry_numerical(&prefix_info, NULL, &gw_addr,
+			GK_FWD_GATEWAY_BACK_NET, update->rt_proto,
+			cps_conf->gk);
+	}
+
+	return -1;
+}
+
+static int
+del_route(struct route_update *update, struct gk_config *gk_conf)
```

Comment:
Style: let's make the prototype of `del_route()` identical to `new_route()`, that is, `del_route(struct route_update *update, struct cps_config *cps_conf)`.